### DISCUSSION TITLE
How to add Home Assistant Climate Entity

### DISCUSSION THREAD

**Discussion by inzinga** [WillCodeForCats/tekmar-482#74] [Q&A] [answered]
Just installed tekMar Gateway 482, with Perle IOLAN SDG2 W device server.  Connected to HA, and both packet server and tekmar-482 in HA seem to be working well.  Image below shows details.  Wondering how to utilize the Home Assistant Climate entity shown at https://github.com/WillCodeForCats/tekmar-482/wiki/Features?

<img width="1061" alt="Screen Shot 2024-03-08 at 11 57 22 AM" src="https://github.com/WillCodeForCats/tekmar-482/assets/8379134/b73964ca-1230-4abe-bb1d-bca82eac25ee">

**Comment by WillCodeForCats** [comment]
That's the gateway, and gateways can't have climate entities.

Climate entities are only on thermostats.

**Reply by inzinga** [reply]
Thermostats are 542's (6, dated Nov 2005, Lot #2202) and 544 (1).  How do I go about adding those climate entities?

**Reply by WillCodeForCats** [reply]
They will show up as devices, like this.

<img width="587" alt="Screenshot 2024-03-08 at 09 19 55" src="https://github.com/WillCodeForCats/tekmar-482/assets/48533968/e263cf67-4921-49a8-bd9e-9499bfed1d43">


<img width="579" alt="Screenshot 2024-03-08 at 09 19 29" src="https://github.com/WillCodeForCats/tekmar-482/assets/48533968/c5b52809-0963-40c1-93d7-ce79a7bee761">

**Reply by WillCodeForCats** [reply]
You can't "add" a climate entity. It is automatic if the thermostat is detected on the bus.

If you do not have any detected thermostats then you can try starting up in debug mode to see why they aren't being detected.

If nothing is detected check the TN bus wiring.

**Reply by inzinga** [reply]
Here's the debug output.  Not seeing anything thermostat related.  Will check bus wiring.  Please let me know if you see any useful information from the log file.

[home-assistant.log](https://github.com/WillCodeForCats/tekmar-482/files/14541709/home-assistant.log)

**Reply by WillCodeForCats** [reply]
The issue appears that the ID during device identification doesn't match what the protocol says it should be. The 542 lists device ID 99302 or the 544 is device ID 99203 in the documentation, but in your logs it shows 10599181. So when it runs setup it ignores unknown devices.

These are the lines in the log that indicate this:
```
2024-03-08 12:33:54.820 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 203. This address will be ignored.
2024-03-08 12:33:55.338 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 1203. This address will be ignored.
```

I don't have a way to override this in config right now, but maybe I can apply a method I recently used in my other integration to manually re-map an unknown ID to a known one if you know the correct device using a YAML config.

The other issue is that you said you have 7 thermostats: the logs show that the gateway is only reading 2 of them. Those you will want to confirm that the TH4 bus is wired.

**Comment by inzinga** [comment]
Okay.  That helps.  I have two 336's, and we only wired up one (which I think is why we're only picking up two thermostats).  Will get the other 336 wired up too.

I have six 542 thermostats (showing device id = 10599181; should be = 99302).

I have one 544 thermostat (showing device id = TBD; should be = 99203).

Will respond after the additional 336 is wired in.

**Reply by WillCodeForCats** [reply]
I'm working on ideas for remapping IDs, but it will take a few days at least.

You can change them in the `const.py` file if you want to try that manually for now.

**Comment by inzinga** [comment]
Copied `const.py` to `const.py.original`, and in original `const.py` copied this block
```
    99302: {                     
        "model": "542",       
        "type": DEVICE_TYPES[99302],
        "heat": 1,            
        "cool": 0,            
        "fan": 0,              
        "humid": 0,            
        "snow": 0,             
        "emer": 0,             
    },                   
```   
and right below pasted back and changed to
```
    10599181: {              
        "model": "542",       
        "type": DEVICE_TYPES[99302],  
        "heat": 1,
        "cool": 0,             
        "fan": 0,         
        "humid": 0,                                                 
        "snow": 0,     
        "emer": 0,                   
    },                     
```
Restarted HA, and still seeing:
```
2024-03-08 22:10:27.767 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 203. This address will be ignored.
2024-03-08 22:10:28.284 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 1203. This address will be ignored.
```

**Reply by WillCodeForCats** [reply]
In `DEVICE_TYPES` add:

```
    10599181: ThaType.THERMOSTAT,
```

**Reply by WillCodeForCats** [reply, marked as answer]
And change the ID at `"type": DEVICE_TYPES[99302],` in your copy paste section to `"type": DEVICE_TYPES[10599181],`

**Comment by inzinga** [comment]
That worked!!!  Thanks.  We're wiring the 2nd 336 tomorrow morning.  Once I find the device id for the single 544, I'll do same procedure (assuming 544 device id not already in `const.py` list).  Will let you know how I make out.  Appreciate the great response from you!

<img width="1254" alt="Screen Shot 2024-03-08 at 10 19 24 PM" src="https://github.com/WillCodeForCats/tekmar-482/assets/8379134/bddd2232-9545-4d5d-a339-b7d4219b9b06">

**Comment by inzinga** [comment]
Also working seamlessly with `HomeKit Bridge`, as I mainly use `HomeKit` to manage my home.  I still need to map each virtual thermostat to the actual physical thermostats, but I'll do that after they're being seen.
![Screenshot 2024-03-08 at 10 29 58 PM](https://github.com/WillCodeForCats/tekmar-482/assets/8379134/19262d5c-ff0c-4005-b9f2-c3ddd00a4b70)
![Screenshot 2024-03-08 at 10 30 36 PM](https://github.com/WillCodeForCats/tekmar-482/assets/8379134/58494ebe-b663-47f2-8d7c-cf5fc62a8803)

**Comment by inzinga** [comment]
Wired in the 2nd bus.  Had all thermos show up except three with device id = `4294967295`.  Repeated process that you had me do yesterday, and restarted HA.  All six 542's and single 544 showing in HA now, with no errors reported in log.

Thanks very, very much for your help with this.

**Comment by iVolt1** [comment]
Old thread, but here I go.  I have two 541 thermostats showing up with the same id as @inzinga so they are ignored by the integration.  I understand the solution provide by @WillCodeForCats but that looks like it would require me to modify the local code or my own fork or risk losing changes on upgrades of the integration.

Is there the possibility of a permanent solution by config or by code for this issue? I don't care too much if the 541s show up as 542s since they are nearly identical anyway.  My five other thermostats display perfectly.

2026-05-28 11:53:19.678 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 201. This address will be ignored.
2026-05-28 11:53:20.387 WARNING (MainThread) [custom_components.tekmar_482.hub] Unknown device type 10599181 at address 1101. This address will be ignored.

Thanks for the great integration.

**Reply by WillCodeForCats** [reply]
I will look at adding it to the list.

**Reply by WillCodeForCats** [reply]
Adding in PR #114
